enderio-base/src/main/java/com/enderio/base/common/capability/darksteel/DarkSteelUpgradeable.java

```diff
@@ -0,0 +1,161 @@
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.nbt.Tag;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.world.item.ItemStack;
+
+import java.util.*;
+
+@SuppressWarnings("unused")
+public class DarkSteelUpgradeable implements IDarkSteelUpgradable {
+
+    //----------------------- Utils
+
+    public static ItemStack addUpgrade(ItemStack itemStack, IDarkSteelUpgrade upgrade) {
+        itemStack.getCapability(EIOCapabilities.DARK_STEEL_UPGRADABLE).ifPresent(upgradable -> upgradable.addUpgrade(upgrade));
+        return itemStack;
+    }
+
+    public static Collection<IDarkSteelUpgrade> getUpgrades(ItemStack itemStack) {
+        return itemStack.getCapability(EIOCapabilities.DARK_STEEL_UPGRADABLE).map(IDarkSteelUpgradable::getUpgrades).orElse(Collections.emptyList());
+    }
+
+    public static boolean hasUpgrade(ItemStack itemStack, String name) {
+        return itemStack.getCapability(EIOCapabilities.DARK_STEEL_UPGRADABLE).map(upgradable -> upgradable.hasUpgrade(name)).orElse(false);
+    }
+
+    public static <T extends IDarkSteelUpgrade> Optional<T> getUpgradeAs(ItemStack itemStack, String upgrade, Class<T> as) {
+        Optional<IDarkSteelUpgradable> cap = itemStack.getCapability(EIOCapabilities.DARK_STEEL_UPGRADABLE).resolve();
+        return cap.flatMap(upgradeCap -> upgradeCap.getUpgradeAs(upgrade, as));
+    }
+
+    public static Collection<IDarkSteelUpgrade> getUpgradesThatCanBeAppliedAtTheMoment(ItemStack itemStack) {
+        return itemStack.getCapability(EIOCapabilities.DARK_STEEL_UPGRADABLE).map(IDarkSteelUpgradable::getUpgradesThatCanBeAppliedAtTheMoment).orElse(Collections.emptyList());
+    }
+
+    public static Collection<IDarkSteelUpgrade> getAllPossibleUpgrades(ItemStack itemStack) {
+        return itemStack.getCapability(EIOCapabilities.DARK_STEEL_UPGRADABLE).map(IDarkSteelUpgradable::getAllPossibleUpgrades).orElse(Collections.emptyList());
+    }
+
+    //----------------------- Class
+
+    private final Map<String, IDarkSteelUpgrade> upgrades = new HashMap<>();
+
+    /** The type of item that is upgradable, used to determine valid upgrades.*/
+    private ResourceLocation onItem;
+
+    public DarkSteelUpgradeable() {
+        this(EnderIO.loc("empty"));
+    }
+
+    public DarkSteelUpgradeable(ResourceLocation onItem) {
+        this.onItem = onItem;
+    }
+
+    @Override
+    public void addUpgrade(IDarkSteelUpgrade upgrade) {
+        removeUpgradeInSlot(upgrade.getSlot());
+        upgrades.put(upgrade.getSerializedName(), upgrade);
+    }
+
+    @Override
+    public void removeUpgrade(String name) {
+        upgrades.remove(name);
+    }
+
+    private void removeUpgradeInSlot(String slot) {
+        Optional<String> toRemove = Optional.empty();
+        for (var entry : upgrades.entrySet()) {
+            if (entry.getValue().getSlot().equals(slot)) {
+                toRemove = Optional.of(entry.getKey());
```

Comment:
Given you break anyways, after you find the first match, wouldn't it make more sense to move the `toRemove.ifPresent` to here and not be creating optionals for basically no reason?

Reply:
Yes, it does.